### ISSUE TITLE
uc_mem_protect() change content if not full region is changed

### ISSUE THREAD

**Issue by PhilippTakacs** [unicorn-engine/unicorn#1549]
If a memory region is mapped over an existing mapping then the protection is changed on a part of the mapping the content of the memory changes.

This looks similar to #1028. 

```
#include "unicorn/unicorn.h"
#include <assert.h>
#include <stdio.h>

#define OK(x) {uc_err __err; if ((__err = x)) { fprintf(stderr, "%s", uc_strerror(__err)); assert(false); } }

static void hook_syscall(uc_engine *uc, void *user_data)
{
	OK(uc_mem_protect(uc, 0x40000000001df000, 0x3000, UC_PROT_READ));
}

static void test_mem_protect()
{
	uc_engine *qc;
	uc_hook hook;
	uint64_t r_rax = 0x40000000001e1bd8;
	uint64_t r_rsi = 0x40000000000096f0;
	uint8_t *data0 = NULL;
	uint8_t *data1 = NULL;
	uint8_t *data2 = NULL;
	uint64_t mem; 
	// mov [rax], rsi
	// xor rsi, rsi
	// syscall
	// mov r14, [rax]
	char code[] = { 0x48, 0x89, 0x30, 0x48, 0x31, 0xF6, 0x0F, 0x05, 0x4C, 0x8B, 0x30 };

	data0 = calloc(sizeof(*data0), 0x1000);
	assert(data0);
	data1 = calloc(sizeof(*data1), 0x1ec000);
	assert(data1);
	data2 = calloc(sizeof(*data2), 0x6000);
	assert(data2);

	OK(uc_open(UC_ARCH_X86, UC_MODE_64, &qc));
	OK(uc_hook_add(qc, &hook, UC_HOOK_INSN, hook_syscall, NULL, 1, 0, UC_X86_INS_SYSCALL));
	OK(uc_reg_write(qc, UC_X86_REG_RAX, &r_rax));
	OK(uc_reg_write(qc, UC_X86_REG_RSI, &r_rsi));

	OK(uc_mem_map_ptr(qc, 0x0, 0x1000, 0x7, data0));
	OK(uc_mem_write(qc, 0x0000, code, sizeof(code)));
	OK(uc_mem_map_ptr(qc, 0x4000000000002000, 0x1ec000, UC_PROT_ALL, data1));
	OK(uc_mem_unmap(qc, 0x40000000001df000, 0x6000));
	OK(uc_mem_map_ptr(qc, 0x40000000001df000, 0x6000, UC_PROT_ALL, data2));

	OK(uc_emu_start(qc, 0x0000, 0x0000 + sizeof(code), 0, 0));
	OK(uc_reg_read(qc, UC_X86_REG_R14, &mem));

	assert(r_rsi == mem);

	OK(uc_close(qc));
}
int main(int argc, char **argv, char **envp)
{
	test_mem_protect();
	return 0;
}
```

**Comment by wtdcode** [comment]
Thanks for your issue and I would have a look ASAP.

However, please expect a bit of delay since I'm quite busy these days.

**Comment by PhilippTakacs** [comment]
I have looked at it and belief I found the bug in split_region():

https://github.com/unicorn-engine/unicorn/blob/141a558dd8b17778e1e12413fd26fb6f2887bb9e/uc.c#L1188-L1195

This stops at the first block where the used_length is >= size of the target memory region. Which is in the example still the mapping at 0x4000000000002000. I don't know how to do this check, because I haven't complete understand the different between a MemoryRegion and a RAMBlock. My guess would be:

```
if (block->mr->addr <= mr->addr && (block->mr->addr + block->used_length) >= mr->end) {
```

This works for the provided test, but I'm not sure if it's correct.

**Comment by wtdcode** [comment]
> I have looked at it and belief I found the bug in split_region():
> 
> https://github.com/unicorn-engine/unicorn/blob/141a558dd8b17778e1e12413fd26fb6f2887bb9e/uc.c#L1188-L1195
> 
> This stops at the first block where the used_length is >= size of the target memory region. Which is in the example still the mapping at 0x4000000000002000. I don't know how to do this check, because I haven't complete understand the different between a MemoryRegion and a RAMBlock. My guess would be:
> 
> ```
> if (block->mr->addr <= mr->addr && (block->mr->addr + block->used_length) >= mr->end) {
> ```
> 
> This works for the provided test, but I'm not sure if it's correct.

Nice catch! I think you are right but I need more tests on this. Thanks!

**Comment by wtdcode** [comment]
Fixed in 96518634fbd2b405af197df15a0d787dd883afa4.

Thanks!
